Emote, say() and visible_message() refactor, tape recorder upgrade

RavingManiac opened the pull request:
get_mobs_in_view() renamed to get_mobs_or_objects_in_view(). Returns mobs or objects depending on specifications

Transmission of emotes, visible_message() and say() refactored. All three use get_mobs_or_objects_in_view() to find objects and mobs in containers

Storage objects and closets no longer transfer hear_talk(), see_emote() or show_message() to contents, as this is already covered by the above changes

visible_message() now affects objects. Holopads, pAIs, and AI cards can now see them

Tape recorders record hearable emotes and action messages

Human and robot emotes now use custom_emote()

RavingManiac committed:
visible_message() now affects objects. Holopads, pAIs, AI cards people in mecha should hence be able to see them

Tape recorders record hearable emotes and action messages

Human and robot emotes now use custom_emote()

mwerezak commented:
While it's a nice idea, `visible_message()` is a pretty high-traffic proc. 

Some performance benchmarks would be good.

GinjaNinja32 commented on code/game/objects/items/weapons/storage/storage.dm:
```diff
@@ -450,6 +450,18 @@
 			O.hear_talk(M, text, verb, speaking)
 
+/obj/item/weapon/storage/see_emote(mob/M as mob, text, var/emote_type)
+	for (var/atom/A in src)
```
`for(var/obj/O in src)` would give you this without the typechecks.

PsiOmegaDelta commented on code/game/objects/items/devices/taperecorder.dm:
```diff
@@ -29,6 +29,25 @@
+	var/recordedtext
+	if (msg && type == 2) //must be hearable
+		recordedtext = msg
+	else if (alt && alt_type == 2)
```
Are there no defines for this magic number?

mwerezak replied:
I think there are, somewhere.

mwerezak replied:
Nope, I was wrong, there are none.

comma commented on code/modules/mob/mob.dm:
```diff
@@ -58,14 +89,40 @@
+
+	var/list/seeing_obj = list()
+
+	for(var/I in view(world.view, get_turf(src))) //get_turf is needed to stop weirdness with x-ray.
```
Can do for var/atom/I in view instead. Won't need any typecasts then, since it has contents, and you don't really need to typecast it to object just to add it to a list. Also won't loop over turfs then.

comma commented on code/modules/mob/mob.dm:
```diff
@@ -58,14 +89,40 @@
+
+	for(var/I in view(world.view, get_turf(src))) //get_turf is needed to stop weirdness with x-ray.
+		if(istype(I, /mob/))
+			var/mob/M = I
```
Why not do the message showing here?
As it is now, you loop over everything once, and THEN loop over mobs in view. Would make sense to do mob stuff here in first loop.

comma commented on code/modules/mob/mob.dm:
```diff
@@ -58,14 +89,40 @@
+		if(istype(I, /mob/))
+			var/mob/M = I
+			for(var/obj/O in M.contents)
+				seeing_obj |= O
```
Maybe do this in mob's show_message and ditch the whole list? 

kinda like this

```
for(var/atom/A in view(world.view,get_turf(src)))
  if(ismob(A))
    var/mob/M = A
    M.show_message(msg,1,blind_message,2) //all message showing to contents should be done there imo
  if(istype(A,/obj))
    var/obj/O = A
    O.show_message( message, 1, blind_message, 2)
```

comma commented on code/modules/mob/mob.dm:
```diff
@@ -51,6 +51,37 @@
+			if(O) //It's possible that it could be deleted in the meantime.
+				O.show_message( message, 1, blind_message, 2)
+
+	/*
```
Why comment it out without removing?

RavingManiac committed:
get_mobs_in_view() renamed to get_mobs_or_objects_in_view(). Returns mobs or objects depending on specifications

Transmission of emotes, visible_message() and say() refactored. All three use get_mobs_or_objects_in_view() to find objects and mobs in containers

Storage objects and closets no longer transfer hear_talk(), see_emote() or show_message() to contents, as this is already covered by the above changes

RavingManiac committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into dev

RavingManiac commented:
New code should be faster for emotes, visible_message() and say()

RavingManiac committed:
Typo fix. Obsolete viisible_message() hack removed from mob holders

RavingManiac committed:
Changelog update

comma commented:
Commited .int

RavingManiac committed:
Reverts .int

Techhead0 commented:
This PR removes the functionality introduced by #9291, namely, the ability to hear the blind_message when invisible people do things.

RavingManiac committed:
Ability to head blind_message when invisible people do things readded

Tape recorder playback uses audible_message instead of visible_message

RavingManiac commented:
Fixed

RavingManiac committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into dev

RavingManiac committed:
Stupid typo

mwerezak commented on code/game/atoms.dm:
```diff
@@ -459,3 +459,46 @@ its easier to just keep the beam vertical.
+// blind_message (optional) is what blind people will hear e.g. "You hear something!"
+/atom/proc/visible_message(var/message, var/blind_message)
+
+	var/list/see = get_mobs_or_objects_in_view(world.view,src) | viewers(get_turf(src), null)
```
I really hope everyone is aware that this makes `visible_message()` call `recursive_content_check()` on *every* atom in view().

mwerezak replied:
Considering how much concern there was in #9408 over searching a single mob when transitioning zlevels - a fairly rare occurrence - I'm surprised no one seems to mind searching every atom in view() every time visible_message() is called.

mwerezak commented on code/game/atoms.dm:
```diff
@@ -459,3 +459,46 @@ its easier to just keep the beam vertical.
+
+	for(var/I in see)
+		if(isobj(I))
+			spawn(0)
```
spawning can be expensive (see - past experiences with the radio controller). Is this really necessary?

mwerezak commented:
You still have not provided performance benchmarks comparing your new code with the old visible_message().

mwerezak commented on code/__HELPERS/game.dm:
```diff
@@ -196,17 +194,17 @@
+				if(M.client)
+					hear += M
+		else if(istype(I,/obj/))
+			hear |= recursive_content_check(I, hear, 3, 1, 0, include_mobs, include_objects)
```
I'm curious why you chose to do a recursive search instead of say, creating a list that objects interested in receiving speech and emotes can add themselves to, and iterating through that?

Ccomp5950 commented:
This never should have been merged.

I'm literally shaking I'm so angry.

PsiOmegaDelta commented:
If things go as I wish it'll be refactored away soon-ish™.
